rust-mode.el

```diff
@@ -323,10 +318,10 @@
   (append
    `(
      ;; Keywords proper
-     (,(regexp-opt rust-mode-keywords 'words) . font-lock-keyword-face)
+     (,(regexp-opt rust-mode-keywords 'symbols) . font-lock-keyword-face)
 
      ;; Special types
-     (,(regexp-opt rust-special-types 'words) . font-lock-type-face)
+     (,(regexp-opt rust-special-types 'symbols) . font-lock-type-face)
```

Comment:
I figured out why the test fails in emacs 23: emacs 23 did not recognize passing `'symbols` to `regexp-opt`, only `'words`.  In emacs 23 this is treated as if you passed `t`, which only groups without surrounding it with `\_<` / `\_>`.  So it starts highlighting keywords in the middle of symbols--for instance it will highlight the "in" as a keyword in `fn main()`.

Replace the above lines with this or something like it, and it should work on all tested versions of emacs:

```
     ;; Keywords proper
     (,(concat "\\_<" (regexp-opt rust-mode-keywords t) "\\_>") . font-lock-keyword-face)

     ;; Special types
     (,(concat "\\_<" (regexp-opt rust-special-types t) "\\_>") . font-lock-type-face)
```